// marvelrivalsapi/lib/api.js
const Achievements = require('./endpoints/achievements');
const Balances = require('./endpoints/balances');
const Battlepass = require('./endpoints/battlepass');
const DevDiaries = require('./endpoints/devdiary');
const Heroes = require('./endpoints/heroes');
const Items = require('./endpoints/items');
const Maps = require('./endpoints/maps');
const Match = require('./endpoints/match');
const MatchHistory = require('./endpoints/matchhistory');
const MatchHistoryv1 = require('./endpoints/matchhistoryv1');
const PatchNotes = require('./endpoints/patchnotes');
const Player = require('./endpoints/player');

class MarvelRivalsAPI {
    constructor(api_key) {
        this.api_key = api_key;

        this.achievements = new Achievements(api_key);
        this.balances = new Balances(api_key);
        this.battlepass = new Battlepass(api_key);
        this.devdiary = new DevDiaries(api_key);
        this.heroes = new Heroes(api_key);
        this.items = new Items(api_key);
        this.maps = new Maps(api_key);
        this.match = new Match(api_key);
        this.matchhistory = new MatchHistory(api_key);
        this.matchhistoryv1 = new MatchHistoryv1(api_key);
        this.patchnotes = new PatchNotes(api_key);
        this.player = new Player(api_key);
    }
}

module.exports = MarvelRivalsAPI;
